File: src/main/java/com/nonxedy/command/NonsceneCommand.java
package com.nonxedy.command;

import com.nonxedy.Nonscenes;
import com.nonxedy.core.ConfigManager;
import com.nonxedy.core.CutsceneManager;
import com.nonxedy.util.ColorUtil;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class NonsceneCommand implements CommandExecutor, TabCompleter {
    private final Nonscenes plugin;
    private final CutsceneManager cutsceneManager;
    private final ConfigManager configManager;

    public NonsceneCommand(Nonscenes plugin) {
        this.plugin = plugin;
        this.cutsceneManager = plugin.getCutsceneManager();
        this.configManager = plugin.getConfigManager();
    }

    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!(sender instanceof Player)) {
            sender.sendMessage(configManager.getMessage("player-only-command"));
            return true;
        }

        Player player = (Player) sender;

        // Check if player has permission
        if (!player.hasPermission("nonscene.use")) {
            player.sendMessage(configManager.getMessage("no-permission"));
            return true;
        }

        if (args.length == 0) {
            sendHelpMessage(player);
            return true;
        }

        String subCommand = args[0].toLowerCase();

        switch (subCommand) {
            case "start":
                if (!player.hasPermission("nonscene.start")) {
                    player.sendMessage(configManager.getMessage("no-permission"));
                    return true;
                }
                
                if (args.length < 3) {
                    player.sendMessage(configManager.getMessage("invalid-start-args"));
                    return true;
                }
                
                String name = args[1];
                int frames;
                
                try {
                    frames = Integer.parseInt(args[2]);
                    if (frames <= 0) {
                        player.sendMessage(configManager.getMessage("invalid-frames-number"));
                        return true;
                    }
                } catch (NumberFormatException e) {
                    player.sendMessage(configManager.getMessage("invalid-frames-number"));
                    return true;
                }
                
                cutsceneManager.startRecording(player, name, frames);
                break;
                
            case "delete":
                if (!player.hasPermission("nonscene.delete")) {
                    player.sendMessage(configManager.getMessage("no-permission"));
                    return true;
                }
                
                if (args.length < 2) {
                    player.sendMessage(configManager.getMessage("specify-cutscene-name"));
                    return true;
                }
                
                cutsceneManager.deleteCutscene(player, args[1]);
                break;
                
            case "all":
                if (!player.hasPermission("nonscene.list")) {
                    player.sendMessage(configManager.getMessage("no-permission"));
                    return true;
                }
                
                cutsceneManager.listAllCutscenes(player);
                break;
                
            case "play":
                if (!player.hasPermission("nonscene.play")) {
                    player.sendMessage(configManager.getMessage("no-permission"));
                    return true;
                }
                
                if (args.length < 2) {
                    player.sendMessage(configManager.getMessage("specify-cutscene-name"));
                    return true;
                }
                
                cutsceneManager.playCutscene(player, args[1]);
                break;
                
            case "showpath":
                if (!player.hasPermission("nonscene.showpath")) {
                    player.sendMessage(configManager.getMessage("no-permission"));
                    return true;
                }
                
                if (args.length < 2) {
                    player.sendMessage(configManager.getMessage("specify-cutscene-name"));
                    return true;
                }
                
                cutsceneManager.showCutscenePath(player, args[1]);
                break;
                
            default:
                sendHelpMessage(player);
                break;
        }
        
        return true;
    }

    private void sendHelpMessage(Player player) {
        for (String line : configManager.getHelpMessages()) {
            player.sendMessage(ColorUtil.format(line));
        }
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
        List<String> completions = new ArrayList<>();
        
        if (!(sender instanceof Player)) {
            return completions;
        }
        
        Player player = (Player) sender;
        
        if (args.length == 1) {
            List<String> subCommands = new ArrayList<>();
            
            if (player.hasPermission("nonscene.start")) subCommands.add("start");
            if (player.hasPermission("nonscene.delete")) subCommands.add("delete");
            if (player.hasPermission("nonscene.list")) subCommands.add("all");
            if (player.hasPermission("nonscene.play")) subCommands.add("play");
            if (player.hasPermission("nonscene.showpath")) subCommands.add("showpath");
            
            return filterCompletions(subCommands, args[0]);
        } else if (args.length == 2) {
            String subCommand = args[0].toLowerCase();
            
            if ((subCommand.equals("delete") || subCommand.equals("play") || subCommand.equals("showpath")) 
                    && player.hasPermission("nonscene." + subCommand)) {
                return filterCompletions(cutsceneManager.getCutsceneNames(), args[1]);
            }
        }
        
        return completions;
    }
    
    private List<String> filterCompletions(List<String> options, String input) {
        return options.stream()
                .filter(option -> option.toLowerCase().startsWith(input.toLowerCase()))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/nonxedy/command/NonsceneCommand.kt
package com.nonxedy.command

import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player
import com.nonxedy.Nonscenes
import com.nonxedy.core.ConfigManager
import com.nonxedy.core.CutsceneManager

class NonsceneCommand(private val plugin: Nonscenes) : CommandExecutor, TabCompleter {
    private val configManager: ConfigManager? = plugin.getConfigManager()
    private val cutsceneManager: CutsceneManager? = plugin.getCutsceneManager()

    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        if (sender !is Player) {
            val message = configManager?.getMessage("player-only-command") ?: "This command can only be used by players."
            sender.sendMessage(message)
            return true
        }

        val player: Player = sender

        // Check if player has permission
        if (!player.hasPermission("nonscene.use")) {
            val message = configManager?.getMessage("no-permission") ?: "You don't have permission to use this command."
            player.sendMessage(message)
            return true
        }

        if (args.isEmpty()) {
            sendHelpMessage(player)
            return true
        }

        val subCommand = args[0].lowercase()

        when (subCommand) {
            "start" -> {
                if (!player.hasPermission("nonscene.start")) {
                    val message = configManager?.getMessage("no-permission") ?: "§cYou don't have permission to use this command."
                    player.sendMessage(message)
                    return true
                }

                if (args.size < 3) {
                    val message = configManager?.getMessage("invalid-start-args") ?: "§cUsage: /nonscene start <name> <frames>"
                    player.sendMessage(message)
                    return true
                }

                val name = args[1]
                val frames: Int

                try {
                    frames = args[2].toInt()
                    if (frames <= 0) {
                        val message = configManager?.getMessage("invalid-frames-number") ?: "§cFrames number must be positive."
                        player.sendMessage(message)
                        return true
                    }
                } catch (e: NumberFormatException) {
                    val message = configManager?.getMessage("invalid-frames-number") ?: "§cInvalid frames number."
                    player.sendMessage(message)
                    return true
                }

                cutsceneManager?.startRecording(player, name, frames)
            }

            "delete" -> {
                if (!player.hasPermission("nonscene.delete")) {
                    val message = configManager?.getMessage("no-permission") ?: "§cYou don't have permission to use this command."
                    player.sendMessage(message)
                    return true
                }

                if (args.size < 2) {
                    val message = configManager?.getMessage("specify-cutscene-name") ?: "§cPlease specify a cutscene name."
                    player.sendMessage(message)
                    return true
                }

                cutsceneManager?.deleteCutscene(player, args[1])
            }

            "all" -> {
                if (!player.hasPermission("nonscene.list")) {
                    val message = configManager?.getMessage("no-permission") ?: "§cYou don't have permission to use this command."
                    player.sendMessage(message)
                    return true
                }

                cutsceneManager?.listAllCutscenes(player)
            }

            "play" -> {
                if (!player.hasPermission("nonscene.play")) {
                    val message = configManager?.getMessage("no-permission") ?: "§cYou don't have permission to use this command."
                    player.sendMessage(message)
                    return true
                }

                if (args.size < 2) {
                    val message = configManager?.getMessage("specify-cutscene-name") ?: "§cPlease specify a cutscene name."
                    player.sendMessage(message)
                    return true
                }

                cutsceneManager?.playCutscene(player, args[1])
            }

            "showpath" -> {
                if (!player.hasPermission("nonscene.showpath")) {
                    val message = configManager?.getMessage("no-permission") ?: "§cYou don't have permission to use this command."
                    player.sendMessage(message)
                    return true
                }

                if (args.size < 2) {
                    val message = configManager?.getMessage("specify-cutscene-name") ?: "§cPlease specify a cutscene name."
                    player.sendMessage(message)
                    return true
                }

                cutsceneManager?.showCutscenePath(player, args[1])
            }

            "stop" -> {
                if (!player.hasPermission("nonscene.stop")) {
                    val message = configManager?.getMessage("no-permission") ?: "§cYou don't have permission to use this command."
                    player.sendMessage(message)
                    return true
                }

                cutsceneManager?.cancelAllSessions(player)
            }

            else -> sendHelpMessage(player)
        }

        return true
    }

    private fun sendHelpMessage(player: Player) {
        val helpMessages = configManager?.getMessageList("help-messages") ?: listOf(
            "§6=== Nonscenes Plugin ===",
            "§7Available commands:",
            "§7- §e/nonscene start <name> <frames> §7- Start recording",
            "§7- §e/nonscene play <name> §7- Play cutscene",
            "§7- §e/nonscene all §7- List cutscenes",
            "§7- §e/nonscene delete <name> §7- Delete cutscene",
            "§7- §e/nonscene showpath <name> §7- Show path",
            "§7- §e/nonscene stop §7- Cancel current action"
        )

        helpMessages.forEach { message ->
            player.sendMessage(message)
        }
    }

    override fun onTabComplete(sender: CommandSender, command: Command, alias: String, args: Array<String>): MutableList<String> {
        val completions = mutableListOf<String>()

        if (sender !is Player) {
            return completions
        }

        val player: Player = sender

        if (args.size == 1) {
            val subCommands = mutableListOf<String>()

            if (player.hasPermission("nonscene.start")) subCommands.add("start")
            if (player.hasPermission("nonscene.delete")) subCommands.add("delete")
            if (player.hasPermission("nonscene.list")) subCommands.add("all")
            if (player.hasPermission("nonscene.play")) subCommands.add("play")
            if (player.hasPermission("nonscene.showpath")) subCommands.add("showpath")
            if (player.hasPermission("nonscene.stop")) subCommands.add("stop")

            return filterCompletions(subCommands, args[0])
        } else if (args.size == 2) {
            val subCommand = args[0].lowercase()

            if ((subCommand == "delete" || subCommand == "play" || subCommand == "showpath")
                && player.hasPermission("nonscene.$subCommand")) {
                return filterCompletions(cutsceneManager?.getCutsceneNames() ?: emptyList(), args[1])
            }
        }

        return completions
    }

    private fun filterCompletions(options: List<String>, input: String): MutableList<String> {
        return options.filter { option ->
            option.lowercase().startsWith(input.lowercase())
        }.toMutableList()
    }
}
